nengo_loihi/neurons.py

```diff
@@ -220,3 +228,98 @@ def nengo_build_nif_rate(model, nif_rate, neurons):
 @nengo.builder.Builder.register(NIF)
 def nengo_build_nif(model, nif, neurons):
     return nengo.builder.neurons.build_lif(model, nif, neurons)
+
+
+if nengo_dl is not None:  # noqa: C901
+    class LoihiLIFBuilder(nengo_dl.neuron_builders.LIFBuilder):
+        def _rate_step(self, J, dt):
+            tau_ref = dt * tf.round(self.tau_ref / dt)
+            tau_ref1 = tau_ref + 0.5*dt
+            # ^ Since LoihiLIF takes `ceil(period/dt)` the firing rate is
+            # always below the LIF rate. Using `tau_ref1` in LIF curve makes
+            # it the average of the LoihiLIF curve (rather than upper bound).
+
+            J -= self.one
+
+            # --- compute Loihi rates (for forward pass)
+            period = tau_ref + self.tau_rc*tf.log1p(tf.reciprocal(
+                tf.maximum(J, self.epsilon)))
+            loihi_rates = (self.amplitude / dt) / tf.ceil(period / dt)
+            loihi_rates = tf.where(J > self.zero, loihi_rates, self.zeros)
+
+            # --- compute LIF rates (for backward pass)
+            if self.config.lif_smoothing:
```

Comment:
Just brainstorming for the future, but it might be interesting to try a version of `loihi_rates` where the steps are smoothed a bit, rather than throwing out the steps altogether.  Same general idea as SoftLIF, but applying smoothing to all the discontinuities in the LoihiLIF function instead of just the threshold.

Reply:
That's an interesting idea. The main reason for the SoftLIF is to smooth the derivative, and here I'm not using the steps at all in the derivative, so the motivation would be a bit different. In practice, though, I think the steps of the tuning curve would get smoothed somewhat in practice due to noise in the network. If you train with noise (which I've been doing for some of the keyword network stuff), then you'll actually get some smoothing of the steps like you're describing, however it will be probabilistic. Doing it deterministically like you suggest would have the benefit of making it more reliable for the network, and therefore allow it to take advantage of it better. You could even (theoretically) use a derivative that takes into account the smoothed steps (maybe that's what you were thinking all along).

The downsides are that you'd have another hyperparameter to tune (you could try to find a good value by estimating the amount of noise on the signals into each layer, and choosing the amount of smoothing to mimic this). The other problem is that I think a simple analytic equation for it (which I think is what we need for TensorFlow) might be difficult to find, keeping in mind that we need it to be relatively easily computable on a GPU. The SoftLIF wasn't trivial to come up with, and that is a considerably simpler problem with only one discontinuity. Applying the same type of SoftLIF smoothing to each of the discontinuities in the stepped Loihi tuning curve would result in an overly-complex function, I think. Maybe there's a way to do it efficiently, though.

Reply:
> You could even (theoretically) use a derivative that takes into account the smoothed steps (maybe that's what you were thinking all along).

Yeah that's what I was thinking.  The nice thing about this implementation is that we're using Loihi-like neurons on the forwards pass, but it would be nice if we could use them on the backwards pass as well.  You're right though that I'm not sure what a compact expression for that would be.